Write disk cache flush failure index=0

From time to time my program using libaria2 aborts with a message saying:

    Write disk cache flush failure index=0

Why does this happen? How can I prevent it from happening? I'm using aria2 version `1.32.0`.

Here's the log:

    [DEBUG] [AbstractSingleDiskAdaptor.cc:101] Cache flush goff=98300, len=16383
    [DEBUG] [AbstractSingleDiskAdaptor.cc:101] Cache flush goff=114683, len=16383
    [ERROR] [WrDiskCacheEntry.cc:83] Error when trying to flush write cache
    Exception: [AbstractDiskWriter.cc:454] errNum=28 errorCode=9 Failed to write into the file /opt/downloads/4a6521b3-b034-407d-aeeb-0bc09dfae682.chunks/2235, cause: No space left on device
    [DEBUG] [AbstractCommand.cc:184] CUID#159304 - socket: read:1, write:0, hup:0, err:0
    [DEBUG] [WrDiskCache.cc:97] Update cache entry size=14232, delta=1120, clock=6049799
    [DEBUG] [Piece.cc:342] updateWrCache entry=0x7ef3ff5179d0
    [DEBUG] [WrDiskCacheEntry.cc:95] WrDiskCacheEntry cache goff=212542, len=15264
    [DEBUG] [WrDiskCache.cc:97] Update cache entry size=29496, delta=15264, clock=6050046
    [DEBUG] [AbstractCommand.cc:184] CUID#159305 - socket: read:1, write:0, hup:0, err:0
    [DEBUG] [Piece.cc:342] updateWrCache entry=0x7ef65fd10fc0
    [DEBUG] [WrDiskCacheEntry.cc:95] WrDiskCacheEntry cache goff=341476, len=3880
    [DEBUG] [WrDiskCache.cc:97] Update cache entry size=113920, delta=3880, clock=6049800
    [INFO] [DownloadCommand.cc:242] CUID#159305 - The download for one segment completed successfully.
    [DEBUG] [WrDiskCache.cc:97] Update cache entry size=113920, delta=-113920, clock=6050048
    [DEBUG] [AbstractSingleDiskAdaptor.cc:101] Cache flush goff=231436, len=16384
    [ERROR] [WrDiskCacheEntry.cc:83] Error when trying to flush write cache
    Exception: [AbstractDiskWriter.cc:454] errNum=28 errorCode=9 Failed to write into the file /opt/downloads/daaf18a1-2368-4dbe-a8e6-5c2088577cc2.chunks/2412, cause: No space left on device
    [DEBUG] [WrDiskCache.cc:97] Update cache entry size=0, delta=0, clock=6050049
    [ERROR] [AbstractCommand.cc:403] CUID#159305 - Download aborted. URI=<reducted>
    Exception: [AbstractCommand.cc:403] errorCode=9 URI=<reducted>
      -> [DownloadCommand.cc:127] errorCode=9 Write disk cache flush failure index=0
    [DEBUG] [RequestGroup.cc:982] GID#b7f7e2c3f9391d6b - Request queue check
    [DEBUG] [AbstractCommand.cc:184] CUID#159307 - socket: read:1, write:0, hup:0, err:0
    [DEBUG] [Piece.cc:342] updateWrCache entry=0x7efc020998e0
    [DEBUG] [WrDiskCacheEntry.cc:95] WrDiskCacheEntry cache goff=256632, len=16383
    [DEBUG] [WrDiskCache.cc:97] Update cache entry size=72231, delta=16383, clock=6049801
    [DEBUG] [AbstractCommand.cc:184] CUID#159309 - socket: read:1, write:0, hup:0, err:0
    [DEBUG] [Piece.cc:342] updateWrCache entry=0x7efa61012930
    [DEBUG] [WrDiskCacheEntry.cc:95] WrDiskCacheEntry cache goff=361017, len=16383
    [DEBUG] [WrDiskCache.cc:972017-08-28 18:05:42.852444 [DEBUG] [RequestGroup.cc:1067] Finding PostDownloadHandler for path /opt/downloads/6952ad89-8bac-472b-9fed-9eb9bc7dd3f8.chunks/2265.
    [DEBUG] [RequestGroup.cc:1080] No PostDownloadHandler found.
    [DEBUG] [RequestGroup.cc:1172] GID#96643b4f21b650d4 - Creating DownloadResult.
    [DEBUG] [DefaultPieceStorage.cc:745] Removed 0 have entries.
    [DEBUG] [ServerStat.cc:111] ServerStat:<reducted>: singleConnectionAvgSpeed_ old:262.10KB/s new:266.41KB/s last:283.65KB/s
    [DEBUG] [WrDiskCache.cc:97] Update cache entry size=0, delta=0, clock=6050050
    [DEBUG] [WrDiskCache.cc:81] Removed cache entry size=0, clock=6050227
    [NOTICE] [RequestGroupMan.cc:427] Download GID#b7f7e2c3f9391d6b not complete: /opt/downloads/daaf18a1-2368-4dbe-a8e6-5c2088577cc2.chunks/2412
    [DEBUG] [RequestGroup.cc:1172] GID#b7f7e2c3f9391d6b - Creating DownloadResult.
    [DEBUG] [DefaultPieceStorage.cc:745] Removed 0 have entries.

It appears that the application is running out of disk space. However, I was under the impression that once there's not enough space to download a file, the download fails with the exit status `9` but won't throw any (undocumented) exception. Is this expected behavior?

--- a/src/aria2api.cc
+++ b/src/aria2api.cc
@@ -722,7 +722,9 @@ struct RequestGroupDH : public DownloadHandle {
   {
     return group->getNumConnection();
   }
-  virtual int getErrorCode() CXX11_OVERRIDE { return 0; }
+  virtual int getErrorCode() CXX11_OVERRIDE {
+    return group->getLastErrorCode();
+  }
   virtual const std::vector<A2Gid>& getFollowedBy() CXX11_OVERRIDE
   {
     return group->followedBy();

